lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -1961,7 +1961,16 @@ static void addOptionalityFixIts(
 /// \brief Diagnose a requirement match, describing what went wrong (or not).
 static void
 diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance,
-              ValueDecl *req, const RequirementMatch &match){
+              ValueDecl *req, const RequirementMatch &match) {
+
+  // If we are matching a constructor and the name doesn't match,
+  // it is likely this witness wasn't intended to be a match at all, so omit
+  // diagnosis.
+  if (req->getKind() == DeclKind::Constructor && match.Kind != MatchKind::RenamedMatch
```

Comment:
I like this for constructors where the problem is most acute, but should we do this for everything?